### PR TITLE
fix: Handle timestamp overflow and fix usec padding on Darwin

### PR DESCRIPTION
**Pull request by kriskowal**
Running a chain on a Mac revealed a couple discrepancies in timestamp handling.

Darwin has different padding and alignment for timeval tv_usec fields, so we saw 0xFFFFFF in the high 32 bits of this 64 bit field. Masking the lower 32 bits compensates. We’re unsure of a better mitigation. It’s likely that this is a discrepancy in stdlib headers.

We also didn’t handle snprintf overflow in our bounds checking.

This addresses both.

### CODE REVIEW

**Comment by kriskowal** [comment]
Paging @phoddie less for the review and more for a clue if there’s a more correct way to extract the microseconds from a timeval portably.

**Comment by mhofman** [comment]
> We also didn’t handle snprintf overflow in our bounds checking.

Well technically it would have been handled had `size` been signed instead of unsigned.

**Review by mhofman — APPROVED** [review]
Fine by me. I still think we should make `size, i, wrote` `int` instead of `unsigned int` as that would prevent underflows in `size` and technically `snprintf` returns an `int`.

**Comment by phoddie** [comment]
Late to the party! I'm not comfortable with the `& 0xFFFFFFFF`. I don't understand why the alignment / padding matters (unless you are sending platform native binary `timeval` structures over the network?). Would you mind explaining a little more to catch me up?

**Comment by kriskowal** [comment]
@phoddie It may simply that my workstation is cursed. Attempting to run a local Agoric chain, xsnap threw an error due to a malformed JSON string, the timing bundle produced on the C side had a closing curly brace in the wrong location.

The full log: https://gist.github.com/kriskowal/d058924896ea52b2727967924b26cdbe

I instrumented xsnap per @mhofman’s suggestion to dump the alleged JSON:

https://gist.github.com/kriskowal/19b264399d548707dae0de41726a9821

This revealed that two things were wrong. The microsecond manitissas were many more than 6 digits long. And the formatter appeared to be continuing despite overrunning the buffer, instead of aborting and returning null.

I verified that the microseconds never differ from each other by more than a million. @mhofman looked at the binary representations and discovered the high half of the 64 bit word was all 1s.

It’s unclear to me how my system is specifically cursed, but I’m betting that my stdlib headers disagree with the platform’s syscall, though I have done nothing special to change my XCode setup. It’s not lexically obvious where xsnap-worker.c gets its definition of the `timeval` struct, so I’m at the end of my very stale, very strictly academic experience with C.

**Comment by mhofman** [comment]
We encountered `tv_usec` that contained values > 1,000,000 on Kris's Mac. When looking closer those, values were in fact of the shape `0xFFFFFFFF000ABCDE`. Apparently Darwin defines `__darwin_suseconds_t` as 32 bit padding and 32 bit value. I'm not sure how that even happened, but apparently the padding ended up as 1s.

**Comment by phoddie** [comment]
Thank you both for catching me up. I'm relieved that these data structures aren't going over the network.

On my Mac, I did a quick test with gettimeofday and timeval.

The `tv_sec` field has an offset of 0 and a size of 8.
The `tv_usec` field has an offset of 8 and a size of 4.

No unexpected padding or alignment. It would be interesting to know what they are for Kris.

Assuming the change to mask is OK (it may be, even if it should be nnecessary), there may be other places in your system that call gettimeofday and don't defend against garbage bits, because they don't expect them - like Date.now(). I suppose in SES that's not used, but... I'm uncomfortable with a mystery about the behavior of such a fundamental function.

**Comment by phoddie** [comment]
A little more... on my Mac the headers are at `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include`. I think these are the relevant bits extracted from there... assuming I read things correctly.

```
typedef int                     __int32_t;

typedef __int32_t       __darwin_suseconds_t;   /* [???] microseconds */

#define _STRUCT_TIMEVAL         struct timeval

_STRUCT_TIMEVAL
{
	__darwin_time_t         tv_sec;         /* seconds */
	__darwin_suseconds_t    tv_usec;        /* and microseconds */
};
#endif /* _STRUCT_TIMEVAL */

```

Using int to define __int32_t makes me nervous, but sizeof(int) is 4, so...

**Comment by phoddie** [comment]
Last thought on this for today.... if somehow Kris' build is using 64-bit ints but macOS expects 32-bit ints, that would explain this. Kris would be allocating 64-bits for tv_usec but macOS would only fill in the low 32-bits. To test that, you could log the size of an `int` and try initializing tv_usec to 0 before calling gettimeofday (if the OS is only writing 32-bits to the 64-bit field, you'll get the high 32-bits consistently as 0 instead of 1 (presumably uninitialized)).

**Comment by kriskowal** [comment]
I have a theory that the problem is actually the sprintf format specifier being "%06lu" regardless of platform, despite differences in the width of the field.

**Comment by kriskowal** [comment]
Or the format specifier should be `%06du` regardless of platform and Linux just happens to leave 0s in the subsequent bytes instead of 1s.

**Comment by phoddie** [comment]
That makes sense. It would be far less troubling too.

**Comment by phoddie** [comment]
Casting the value to ensure it is consistent across platforms and matches the format specifier seems correct.

File: xsnap/sources/xsnap-worker.c [hunk 1]
```diff
@@ -205,8 +205,18 @@ static char *renderTimestamps() {
...
+		// the expected number of bits behind the variadic args reference.
+		// We do the same for tv_usec out of an outrageous abundance of
+		// caution.
 		wrote = snprintf(p, size, "%lu.%06lu",
-						 timestamps[i].tv_sec, timestamps[i].tv_usec);
+						 (unsigned long)timestamps[i].tv_sec,
+						 (unsigned long)timestamps[i].tv_usec);
+		if (wrote > size) {
+			return NULL;
+		}
```
Comment (mhofman) [inline xsnap/sources/xsnap-worker.c:232]: Technically redundant now we have size as signed, but it likely doesn't cost much too keep this in.

**Review by mhofman — APPROVED** [review]
Thanks for digging for the best way to solve this.

File: xsnap/sources/xsnap-worker.c (hunk 1 above)
Comment (kriskowal) [reply xsnap/sources/xsnap-worker.c:232]: Yeah, it’s almost sufficiently paranoid.

**Comment by kriskowal** [comment]
I’m holding this back until I get a green integration build. https://github.com/Agoric/agoric-sdk/pull/7885

**Comment by kriskowal** [comment]
Integration build is green. https://github.com/Agoric/agoric-sdk/pull/7885

**Pull request merged**